### PR TITLE
Pipeline v5

### PR DESCRIPTION
**Pull request by mberacochea**

### CODE REVIEW

File: emgapianns/management/commands/import_contigs.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                            help='KEGG Modules summary file.')
+        parser.add_argument('--min-length', action='store', type=int, default=500,
+                            help='Only import contigs longer that this value.', required=False)
+
+    def populate_from_accession(self, options):
+        logger.info('Found {}'.format(len(self.obj_list)))
+        for analysis_job in self.obj_list:
+            self.load_contigs(analysis_job, options)
+
+    def load_contigs(self, analysis_job, options):  # noqa: C901
```
Comment (mscheremetjew) [inline emgapianns/management/commands/import_contigs.py]: This method is far to long.

File: emgapianns/management/commands/import_contigs.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        gff = options['gff']
+        kegg_modules = options['kegg_modules']
+        antismash = options['antismash']
+        min_length = options['min_length']
+        batch_size = options['batch_size']
+
+        logger.info('Starting the contigs import process for: ' + str(analysis_job.accession))
+
+        # load the gff in memory
+        annotations_dict = {}
```
Comment (mscheremetjew) [inline emgapianns/management/commands/import_contigs.py]: The GFF load could be factored out into a separate method.

File: emgapianns/management/commands/import_contigs.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                        'interpro': [],
+                        'go': []
+                    }
+                for category in atts.split(';'):
+                    for possible_cat in ['kegg', 'cog', 'pfam', 'interpro', 'go']:
+                        if category.startswith(possible_cat + '='):
+                            values = Command._split(category.replace(possible_cat + '=', ''))
+                            annotations_dict[contig_id][possible_cat].extend(values)
+
+        if antismash:
```
Comment (mscheremetjew) [inline emgapianns/management/commands/import_contigs.py]: This one could be moved into its own method too.

File: emgapianns/management/lib/__init__.py [hunk 4]
```diff
@@ -29,8 +32,10 @@ def add_arguments(self, parser):
         )
         parser.add_argument(
             '--pipeline',
+            help='Pipeline version',
             action='store',
-            dest='pipeline'
+            dest='pipeline',
+            choices=['4.1', '5.0'], default='4.1'
```
Comment (mscheremetjew) [inline emgapianns/management/lib/__init__.py]: Should we change the default value to 5.0?

Comment (mscheremetjew) [inline emgapianns/management/lib/__init__.py]:
```suggestion
            choices=['4.1', '5.0'], default='5.0'
```

File: emgapianns/mixins.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                    .get(analysis_id=str(job.job_id))
+        except self.annotation_model.DoesNotExist:
+            # Return an empty list, the entity exists
+            # but it doesn't have annotations
+            return []
+
+        return getattr(analysis, self.annotation_model_property, [])
+
+
+class AnnotationRetrivalMixin():
```
Comment (mscheremetjew) [inline emgapianns/mixins.py]: Typo. Please change to AnnotationRetrievalMixin

File: emgapianns/mixins.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    """FIXME
```
Comment (mscheremetjew) [inline emgapianns/mixins.py]: I assume the FIXME can go?

File: emgapianns/views.py [hunk 7]
```diff
@@ -87,30 +88,17 @@ def retrieve(self, request, *args, **kwargs):
             .retrieve(request, *args, **kwargs)
 
 
-class InterproIdentifierViewSet(m_viewsets.ReadOnlyModelViewSet):
-
+class InterproIdentifierViewSet(m_mixins.AnnotationRetrivalMixin, m_viewsets.ReadOnlyModelViewSet):
```
Comment (mscheremetjew) [inline emgapianns/views.py]: Typo in AnnotationRetrivalMixin. Change to Retrieval

File: emgapianns/views.py [hunk 8]
```diff
@@ -135,235 +123,456 @@ def retrieve(self, request, *args, **kwargs):
             .retrieve(request, *args, **kwargs)
 
 
-class GoTermAnalysisRelationshipViewSet(m_viewsets.ListReadOnlyModelViewSet):
+class KeggModuleViewSet(m_mixins.AnnotationRetrivalMixin, m_viewsets.ReadOnlyModelViewSet):
```
Comment (mscheremetjew) [inline emgapianns/views.py]: Typo. Change to AnnotationRetrievalMixin

File: emgapianns/views.py [hunk 9]
```diff
@@ -135,235 +123,456 @@ def retrieve(self, request, *args, **kwargs):
...
+        """
+        return super(KeggModuleViewSet, self) \
+            .retrieve(request, *args, **kwargs)
 
-    search_fields = (
-        '@sample__metadata__var_val_ucv',
-    )
 
-    lookup_field = 'accession'
+class PfamEntryViewSet(m_mixins.AnnotationRetrivalMixin, m_viewsets.ReadOnlyModelViewSet):
```
Comment (mscheremetjew) [inline emgapianns/views.py]: typo: Change to AnnotationRetrievalMixin

File: emgapianns/views.py [hunk 10]
```diff
@@ -476,103 +689,158 @@ class AnalysisOrganismRelationshipViewSet(  # NOQA
...
 
-        return super(AnalysisOrganismLSURelationshipViewSet, self) \
-            .list(request, *args, **kwargs)
+        return Response({
+            'accession': analysis.accession,
+            'pipeline_version': analysis.pipeline_version,
+            'taxonomy_count': len(getattr(analysis, 'taxonomy', [])),
+            'taxonomy_ssu_count': len(getattr(analysis, 'taxonomy_ssu', [])),
+            'taxonomy_lsu_count': len(getattr(analysis, 'taxonomy_lsu', [])),
+            'taxonomy_itsunite_count': len(getattr(analysis, 'taxonomy_itsunite', [])),
```
Comment (mscheremetjew) [inline emgapianns/views.py]: Is this a reference to the directory name on the file system? It not we can leave it like that.

File: nginx/conf.d/emgapi.conf [hunk 11]
```diff
@@ -1,48 +0,0 @@
-ignore_invalid_headers off;
```
Comment (mscheremetjew) [inline nginx/conf.d/emgapi.conf]: Has this file been replaced or did it become obsolete.

File: tests/integration/test_annotations.py [hunk 12]
```diff
@@ -22,21 +22,40 @@
...
 
 @pytest.mark.usefixtures('mongodb')
 @pytest.mark.django_db
-class TestAnnotations(object):
+class TestAnnotations:
+
+    @property
+    def _annotation_endpoints(self):
+        return [
+            ['emgapi_v1:analysis-goslim-list', 'go_slim'],
```
Comment (mscheremetjew) [inline tests/integration/test_annotations.py]: Good to see that you add tests for all the new features.

File: travis/download_description_label.sql [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+LOCK TABLES `DOWNLOAD_DESCRIPTION_LABEL` WRITE;
+/*!40000 ALTER TABLE `DOWNLOAD_DESCRIPTION_LABEL` DISABLE KEYS */;
+INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('MAPSeq output file for SSU','MAPseq SSU assignments');
+INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('MAPSeq output file for LSU','MAPseq LSU assignments');
+INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('Processed contigs','Processed contigs');
+INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('All contigs that have predicted CDS','Processed contigs with pCDS');
+INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('All contigs with an interproscan match','Processed contigs with annotation');
+INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('All contigs with a predicted CDS but no interproscan match','Processed contigs without annotation');
+INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('All contigs encoding 5S rRNA','Contigs encoding 5S rRNA');
```
Comment (mscheremetjew) [inline travis/download_description_label.sql]:
```suggestion
INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('All contigs encoding 5S rRNA','Contigs encoding 5S rRNA');
INSERT INTO `DOWNLOAD_DESCRIPTION_LABEL` (DESCRIPTION, DESCRIPTION_LABEL) VALUES ('All contigs encoding 5.8S rRNA','Contigs encoding 5.8S rRNA');
```

File: emgapianns/views.py (hunk 10 above)
Comment (mberacochea) [reply emgapianns/views.py]: No, that is on the model.

File: tests/integration/test_annotations.py (hunk 12 above)
Comment (mberacochea) [reply tests/integration/test_annotations.py]: We need to add more. :)

File: nginx/conf.d/emgapi.conf (hunk 11 above)
Comment (mberacochea) [reply nginx/conf.d/emgapi.conf]: that is local dev stuff, I'll add JIRA task to clean stuff.

**Pull request merged**